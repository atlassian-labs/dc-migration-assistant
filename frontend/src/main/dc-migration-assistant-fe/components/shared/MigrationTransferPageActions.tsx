/*
 * Copyright 2020 Atlassian
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { FunctionComponent } from 'react';
import Button from '@atlaskit/button';
import { Link } from 'react-router-dom';
import { CancelButton } from './CancelButton';

export type MigrationTransferActionsProps = {
    /**
     * A function which will be called when the Refresh button is clicked
     */
    onRefresh: () => Promise<void>;
    /**
     * Whether the current migration transfer has completed or not
     */
    finished: boolean;
    /**
     * The text to display on the button that takes the user to the next step of the migration
     */
    nextText: string;
    /**
     * The path to take the user to whent hey click the "next" action button
     */
    nextRoute: string;
    /**
     * A function which will be called when the "start" action button is called.
     */
    startMigrationPhase: () => Promise<void>;
    /**
     * Whether the current migration transfer has already started or not
     */
    started: boolean;
    /**
     * Should be true if any of the above properties are being fetched
     */
    loading: boolean;
};

/**
 * A component which renders the actions that can be taken at a given step of the migration.
 * The "Cancel" action is always available and cancels the current step. There are three other
 * possible actions which can be availabl depending on the props:
 *   **Start**: Will be available when the current transfer hasn't started
 *   **Refresh**: Will be available when the current transfer is in progress
 *   **Next**: Will be available when the current transfer is completed. It's text can be overriden
 */
export const MigrationTransferActions: FunctionComponent<MigrationTransferActionsProps> = ({
    finished,
    nextText,
    nextRoute,
    startMigrationPhase,
    onRefresh: updateTransferProgress,
    started,
    loading,
}) => {
    const defaultButtonStyle = {
        padding: '5px',
    };

    const marginButtonStyle = {
        ...defaultButtonStyle,
        marginRight: '5px',
    };

    const StartButton = (
        <Button
            style={marginButtonStyle}
            isLoading={loading}
            appearance="primary"
            onClick={startMigrationPhase}
        >
            Start
        </Button>
    );

    let ActionButton = StartButton;

    if (finished) {
        const NextButton = (
            <Link to={nextRoute}>
                <Button style={defaultButtonStyle} appearance="primary">
                    {nextText}
                </Button>
            </Link>
        );
        ActionButton = NextButton;
    } else if (started) {
        const RefreshButton = (
            <Button style={marginButtonStyle} isLoading={loading} onClick={updateTransferProgress}>
                Refresh
            </Button>
        );
        ActionButton = RefreshButton;
    }

    return (
        <>
            {ActionButton}
            <CancelButton />
        </>
    );
};
