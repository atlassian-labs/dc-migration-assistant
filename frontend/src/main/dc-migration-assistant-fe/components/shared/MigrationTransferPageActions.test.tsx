/*
 * Copyright 2020 Atlassian
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter as Router } from 'react-router-dom';
import { MigrationTransferActions } from './MigrationTransferPageActions';

describe('Migration step action', () => {
    const startFunction = jest.fn();
    const startButtonText = 'Start';
    const StartButton = (
        <Router>
            <MigrationTransferActions
                state="not_started"
                onRefresh={(): Promise<void> => Promise.resolve()}
                nextText="Next"
                nextRoute="/"
                startButtonText={startButtonText}
                startMigrationPhase={startFunction}
                loading={false}
            />
        </Router>
    );

    it('should render start button when migration step is not started', () => {
        const { getByText } = render(StartButton);

        expect(getByText(startButtonText)).toBeTruthy();
    });

    it('should call the start function when start button is clicked', () => {
        const { getByText } = render(StartButton);

        getByText(startButtonText).click();

        expect(startFunction).toBeCalled();
    });

    const refreshFunction = jest.fn();
    const refreshButtonText = 'atlassian.migration.datacenter.generic.refresh';
    const RefreshButton = (
        <Router>
            <MigrationTransferActions
                state="in_progress"
                onRefresh={refreshFunction}
                nextText="Next"
                nextRoute="/"
                startButtonText="Start"
                startMigrationPhase={(): Promise<void> => Promise.resolve()}
                loading={false}
            />
        </Router>
    );
    it('should render refresh button when migration step is in progress', () => {
        const { getByText } = render(RefreshButton);

        expect(getByText(refreshButtonText)).toBeTruthy();
    });

    it('should call the refresh function when the refresh button is clicked', () => {
        const { getByText } = render(RefreshButton);

        getByText(refreshButtonText).click();

        expect(refreshFunction).toBeCalled();
    });

    const nextRoute = '/next';
    const nextButtonText = 'Next';
    const NextButton = (
        <Router>
            <MigrationTransferActions
                state="finished"
                onRefresh={(): Promise<void> => Promise.resolve()}
                nextText={nextButtonText}
                nextRoute={nextRoute}
                startButtonText="Start"
                startMigrationPhase={(): Promise<void> => Promise.resolve()}
                loading={false}
            />
        </Router>
    );

    it('should render the next button when migration step is finished', () => {
        const { getByText } = render(NextButton);

        expect(getByText(nextButtonText)).toBeTruthy();
    });

    it('should redirect to the given next route when clicking next', () => {
        // TODO
    });
});
