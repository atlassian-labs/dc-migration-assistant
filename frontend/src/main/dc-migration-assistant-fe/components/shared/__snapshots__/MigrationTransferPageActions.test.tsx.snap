// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migration step action should render a spinner button when loading 1`] = `
<button
  class="css-1wbvwgu"
  data-testid="button-start"
  style="padding: 5px;"
  type="button"
>
  <span
    class="css-j8fq0c"
  >
    <div
      class="css-d0s4f5"
    >
      <svg
        class="css-wzerkf"
        focusable="false"
        height="24"
        viewBox="0 0 16 16"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          class="css-4mzqsf"
          cx="8"
          cy="8"
          r="7"
        />
      </svg>
    </div>
    <span
      class="css-ysfjrh"
    >
      Start
    </span>
  </span>
</button>
`;
